File: s_coding/11724_연결 요소의 개수.cpp
#if 01
//연결 요소의 개수
//https://www.acmicpc.net/problem/11724
#include <iostream>
using namespace std;
int N, M;// 정점의 개수, 간선의 개수 
int adj[1000 + 10][1000 + 10];
int visited[1000 + 10];

void DFS(int v) {
	visited[v] = 1;

	for (int i = 1; i <= N; i++) {
		if (visited[i] == 1 || adj[v][i] == 0 || adj[i][v] == 0)continue;
		DFS(i);
	}
}
int main() {
	int cnt = 0;

	freopen("input.txt", "r", stdin);
	cin >> N >> M;

	for (int i = 0; i <M; i++) {
		int a, b;
		cin >> a >> b;
		adj[a][b] = adj[b][a] = 1;
	}

	for (int i = 1; i <= N; i++) {
		if (visited[i] == 0) {
			DFS(i);
			cnt++;
		}
	}

	cout << cnt;
	return 0;
}
#endif

File: s_coding/BOJ/11724_연결 요소의 개수.py
#연결 요소의 개수
#https://www.acmicpc.net/problem/11724
import sys
sys.stdin = open("input.txt","r")
input=sys.stdin.readline
sys.setrecursionlimit(10**8)

def DFS(v):
    #v가 시작이 됨
    visited[v]=1

    for i in range(1,n+1):
        #둘다 연결되어 있고 아직 방문안했을 때
        if board[v][i]==1 and board[i][v]==1 and visited[i]==0 :
            DFS(i)


if __name__=="__main__":
    n, m = map(int, input().split())
    board = [[0] * (n + 1) for _ in range(n + 1)]
    visited=[0]*(n+1)
    for _ in range(m):
        s, e = map(int, input().split())
        board[s][e] = board[e][s] = 1

    cnt=0

    for i in range(1,n+1):
            if visited[i]==0: #방문아직 안했으면
                DFS(i)
                cnt+=1
    print(cnt)
